Common sql bugfixes and improvements

denimalpaca opened the pull request:
Large PR that will fix a few bug issues, rework check operators, and add some new functionality.
List of changes:
- Fix a test for the BigQueryTableCheckOperator
- Remove unnecessary job_id generation in BigQuery check operators
- Rework SQL building and move to `init()` from `execute()`
- Rework SQLColumnCheckOperator to send a single query with a UNION ALL of checks for each column
- Add a check-level partition statement to allow greater functionality in each check
- Add an option for an AirflowFailException, which will short-circuit retries on failure #26712 

closes: #26712
related: #26368

Commit:
Fix BigQueryTableCheckOperator test

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Fix BigQueryTableCheckOperator test

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Fix BigQueryTableCheckOperator test

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Remove job_id generation in table/col check operators

The job_id is automatically generated by hook.insert_job()
if an empty string is passed, so job_id generation in the
operator is removed in favor of the existing code.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Remove job_id generation in table/col check operators

The job_id is automatically generated by hook.insert_job()
if an empty string is passed, so job_id generation in the
operator is removed in favor of the existing code.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Remove job_id generation in table/col check operators

The job_id is automatically generated by hook.insert_job()
if an empty string is passed, so job_id generation in the
operator is removed in favor of the existing code.

Signed-off-by: Benji Lampel <benjamin@astronomer.io>

Commit:
Rework SQL query building

SQL query building is moved to the init() method of the column
and table check operators to lessen the amount of duplicate code
in the child operator. It also has the added effect of, ideally,
passing a more complete query to OpenLineage.

In doing the above, the column check operator had to be reworked and
now matches the logic of the table check operator in terms of
returning multiple rows and only sending one query to the database.

Commit:
Rework SQL query building

SQL query building is moved to the init() method of the column
and table check operators to lessen the amount of duplicate code
in the child operator. It also has the added effect of, ideally,
passing a more complete query to OpenLineage.

In doing the above, the column check operator had to be reworked and
now matches the logic of the table check operator in terms of
returning multiple rows and only sending one query to the database.

Commit:
Rework SQL query building

SQL query building is moved to the init() method of the column
and table check operators to lessen the amount of duplicate code
in the child operator. It also has the added effect of, ideally,
passing a more complete query to OpenLineage.

In doing the above, the column check operator had to be reworked and
now matches the logic of the table check operator in terms of
returning multiple rows and only sending one query to the database.

Commit:
Remove self.sql overwrite parameter in BigQuery check operators.

Commit:
Remove self.sql overwrite parameter in BigQuery check operators.

Commit:
Remove self.sql overwrite parameter in BigQuery check operators.

Commit:
Add option to fail check operators without retries

Adds a new parameter, retry_on_failure, and a new function to
determine if operators should retry or not on test failure.

Commit:
Add option to fail check operators without retries

Adds a new parameter, retry_on_failure, and a new function to
determine if operators should retry or not on test failure.

Commit:
Add option to fail check operators without retries

Adds a new parameter, retry_on_failure, and a new function to
determine if operators should retry or not on test failure.

Commit:
Rename and reorder private functions, fix typo

Commit:
Rename and reorder private functions, fix typo

Commit:
Rename and reorder private functions, fix typo

bhirsz commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -60,6 +74,12 @@ def _get_failed_checks(checks, col=None):
 
+def _raise_exception(exception_string, retry_on_failure):
+    if retry_on_failure:
+        raise AirflowException(exception_string)
```
This method raises the same exception, with or without retry_on_failure being True

denimalpaca replied:
This method comes from this [issue](https://github.com/apache/airflow/discussions/26712#discussioncomment-3747662); the `AirflowFailException` should halt the task from retrying.

bhirsz commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -33,7 +33,21 @@
-def parse_boolean(val: str) -> str | bool:
+def _convert_to_float_if_possible(s):
+    """
+    A small helper function to convert a string to a numeric value
```
The code is simple and self explanatory, I think we can skip docs

uranusjr replied:
Type hints would however be a good idea here

bhirsz commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -33,7 +33,21 @@
+    :param s: the string to be converted
+    """
+    try:
+        ret = float(s)
```
Unnecessary variable, you can return directly (return float(s) or return s in except)

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -60,6 +74,12 @@ def _get_failed_checks(checks, col=None):
     ]
 
 
+def _raise_exception(exception_string, retry_on_failure):
```
```suggestion
def _raise_exception(exception_string: str, retry_on_failure: bool) -> NoReturn:
```

NoReturn should be imported from typing.

Commit:
Update airflow/providers/common/sql/operators/sql.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Commit:
Update airflow/providers/common/sql/operators/sql.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Commit:
Update airflow/providers/common/sql/operators/sql.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Commit:
Small updates to helper functions

Commit:
Small updates to helper functions

Commit:
Small updates to helper functions

Commit:
Use _raise_exception when query returns 0 rows

Commit:
Use _raise_exception when query returns 0 rows

Commit:
Use _raise_exception when query returns 0 rows

Commit:
Update tests and operator code

Updates tests to reflect changes in operator code, and fixed
bugs in operators as well. Mainly moving the code to check for
failed tests into the column and table check operators as it works
slightly differently for each and doesn't make much sense as a
top-level function.

Commit:
Update tests and operator code

Updates tests to reflect changes in operator code, and fixed
bugs in operators as well. Mainly moving the code to check for
failed tests into the column and table check operators as it works
slightly differently for each and doesn't make much sense as a
top-level function.

Commit:
Update tests and operator code

Updates tests to reflect changes in operator code, and fixed
bugs in operators as well. Mainly moving the code to check for
failed tests into the column and table check operators as it works
slightly differently for each and doesn't make much sense as a
top-level function.

Commit:
Fix _failed_checks() to match update in parent operators

Commit:
Fix _failed_checks() to match update in parent operators

Commit:
Fix _failed_checks() to match update in parent operators

Commit:
Insert quotes around check_name table op's sql template to fix sql error

Commit:
Insert quotes around check_name table op's sql template to fix sql error

Commit:
Insert quotes around check_name table op's sql template to fix sql error

Commit:
Merge branch 'common_sql_bugfixes_and_improvements' of github.com:denimalpaca/airflow into common_sql_bugfixes_and_improvements

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -229,46 +235,65 @@ def __init__(
+        for column, checks in self.column_mapping.items():
+            for check, check_values in checks.items():
+                self._column_mapping_validation(check, check_values)
+            checks_list = [*checks]
```
This doesn’t seem to be needed? As far as I can tell you can simply do `for check in checks` beflow.

uranusjr reviewed (approved):
I’d rewrite all the string-building logic into generators, but the logic this is trying to implement looks good, and performance is not really critical in any of this anyway.

Commit:
Remove unnecessary list comprehension

Commit:
Remove unnecessary list comprehension

Commit:
Added assertions in existing column and table tests

Commit:
Added assertions in existing column and table tests

Commit:
Add new tests for TableCheckOperator

Commit:
Add new tests for TableCheckOperator

Commit:
Update operator logic and tests

Adds "where" option in checks dictionaries for column and table
operators, which may be renamed. This allows for check-level
partitioning, whereas the partition_clause param will always be
for all checks. New tests are added for this addition.

Commit:
Update operator logic and tests

Adds "where" option in checks dictionaries for column and table
operators, which may be renamed. This allows for check-level
partitioning, whereas the partition_clause param will always be
for all checks. New tests are added for this addition.

Commit:
Merge branch 'apache:main' into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'common_sql_bugfixes_and_improvements' of github.com:denimalpaca/airflow into common_sql_bugfixes_and_improvements

Commit:
Add testing for column check operator and line edits

Cleans up operator and adds testing for new generator function.

Commit:
Add testing for column check operator and line edits

Cleans up operator and adds testing for new generator function.

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Change name 'where' to 'partition_clause' in check dictionaries

Commit:
Change name 'where' to 'partition_clause' in check dictionaries

Commit:
Update docs and use f strings

Commit:
Update docs and use f strings

Commit:
Edit operator docstring

Commit:
Edit operator docstring

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -229,49 +237,81 @@ def __init__(
-            for idx, result in enumerate(records):
-                tolerance = self.column_mapping[column][checks[idx]].get("tolerance")
+        for row in records:
+            column, check, result = row
```
```suggestion
        for column, check, result in records:
```

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -229,49 +237,81 @@ def __init__(
+                    for check, check_values in checks.items()
+                    if not check_values["success"]
+                ]
             )
```
```suggestion
        failed_tests = [
            f"Column: {col}\n\tCheck: {check},\n\tCheck Values: {check_values}\n"
            for col, checks in self.column_mapping.items()
            for check, check_values in checks.items()
            if not check_values["success"]
        ]
```

And the `failed_tests = []` line above can be removed.

denimalpaca replied:
Didn't know you could do multiple loops in list comprehensions!

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -229,49 +237,81 @@ def __init__(
+                )
+                for check in checks
+            ]
+        )
```
```suggestion
        checks_sql = "UNION ALL".join(
            self.sql_check_template.format(
                check_statement=self.column_checks[check].format(column=column),
                check=check,
                table=self.table,
                column=column,
                partition_clause=_generate_partition_clause(check),
            )
            for check in checks
        )
```

`join` takes a generator expression so the extra `[]` is not necessary (and they slow things down!)

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -418,46 +461,59 @@ def __init__(
+                Test failed.\nQuery:\n{self.sql}\nResults:\n{records!s}\n
+                The following tests have failed:
+                \n{', '.join(failed_tests)}
+            """
```
```suggestion
            exception_string = (
                f"Test failed.\nQuery:\n{self.sql}\nResults:\n{records!s}\n"
                f"The following tests have failed:\n{', '.join(failed_tests)}"
            )
```

Otherwise the exception message would contain weird indents

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -46,18 +53,10 @@ def parse_boolean(val: str) -> str | bool:
+def _raise_exception(exception_string: str, retry_on_failure: bool) -> NoReturn:
+    if retry_on_failure:
+        raise AirflowException(exception_string)
+    raise AirflowFailException(exception_string)
```
Since this function is always used with `self.retry_on_failure`, perhaps it should be a method instead and we can get rid of the second argument?

uranusjr reviewed (commented):
The general structure looks good to me, just some stylish comments. Most of them are repated in multiple operators so it’d be a good idea to change them all… or maybe extra the logic in a base class? Not sure about that.

Commit:
Move _raise_exception to base class to simplify method

Commit:
Move _raise_exception to base class to simplify method

Commit:
Updates from code review

Commit:
Updates from code review

Commit:
Merge branch 'common_sql_bugfixes_and_improvements' of github.com:denimalpaca/airflow into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Update BigQuery Check operators according to code review

Commit:
Update BigQuery Check operators according to code review

uranusjr commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -229,49 +236,75 @@ def __init__(
+            for check, check_values in checks.items():
+                self._column_mapping_validation(check, check_values)
+            checks_sql_list.append(self._generate_sql_query(column, checks))
+        checks_sql = "UNION ALL".join(checks_sql_list)
```
```suggestion
        def _build_checks_sql():
            for column, checks in self.column_mapping.items():
                for check, check_values in checks.items():
                    self._column_mapping_validation(check, check_values)
                yield self._generate_sql_query(column, checks)

        checks_sql = "UNION ALL".join(_build_checks_sql())
```

uranusjr committed:
Rewrite data-building loop to generator

uranusjr committed:
Rewrite data-building loop to generator

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Add new accept_none argument to column check operator

The new argument, defaulting to true, will convert Nones returned
from the query to 0s so numeric calculations can be performed
correctly. This allows empty tables to be handled as a row of zeroes.

Additional documentation is also supplied

Commit:
Add new accept_none argument to column check operator

The new argument, defaulting to true, will convert Nones returned
from the query to 0s so numeric calculations can be performed
correctly. This allows empty tables to be handled as a row of zeroes.

Additional documentation is also supplied

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

bhirsz reviewed (approved):
LGTM

Commit:
Fix formatting issues after rebase

Commit:
Merge branch 'main' into common_sql_bugfixes_and_improvements

Commit:
Merge branch 'common_sql_bugfixes_and_improvements' of github.com:denimalpaca/airflow into common_sql_bugfixes_and_improvements
